## Lab 6
## Author:Samikshya Pandey

In [1]:
#Calling pandas and numpy 
import pandas as pd
import numpy as np

In [2]:
#Reading the files to the value
netflix_title = pd.read_csv("netflix_titles.csv")
netflix_title.head()
netflix_title.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8807 entries, 0 to 8806
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   show_id       8807 non-null   object
 1   type          8807 non-null   object
 2   title         8807 non-null   object
 3   director      6173 non-null   object
 4   cast          7982 non-null   object
 5   country       7976 non-null   object
 6   date_added    8797 non-null   object
 7   release_year  8807 non-null   int64 
 8   rating        8803 non-null   object
 9   duration      8804 non-null   object
 10  listed_in     8807 non-null   object
 11  description   8807 non-null   object
dtypes: int64(1), object(11)
memory usage: 825.8+ KB


### Question 1 


In [3]:
#Finding number of missing values in each column 

netflix_title.isna().sum()

show_id            0
type               0
title              0
director        2634
cast             825
country          831
date_added        10
release_year       0
rating             4
duration           3
listed_in          0
description        0
dtype: int64

There are 2634 missing values for director, 825 for cast and 831 for country. Rating has 4 missing value and duration has 3 missing values. 


### Question 2
Remove the rows which have empty value for the column “director"

In [4]:
netflix_title = netflix_title.dropna(subset =['director'])
netflix_title.isna().sum()  #confirming to see that director has no missing value as we dropped as na values

show_id           0
type              0
title             0
director          0
cast            473
country         422
date_added        0
release_year      0
rating            1
duration          3
listed_in         0
description       0
dtype: int64

### Question 3
Replace the missing value in the column “ratings” with the mode of the column’s values.


In [5]:
#netflix_title.head()

In [6]:
print(netflix_title["rating"].unique())
print("-----")
#Looking at the unique values. Counting the values within the column.
#Mode is the most frequent value so value count lists us the mode value of the rating column
print(netflix_title["rating"].value_counts() )

['PG-13' 'TV-MA' 'PG' 'TV-14' 'TV-PG' 'TV-Y' 'R' 'TV-G' 'TV-Y7' 'G'
 'NC-17' '74 min' '84 min' '66 min' 'NR' 'TV-Y7-FV' 'UR' nan]
-----
TV-MA       2117
TV-14       1457
R            795
TV-PG        538
PG-13        489
PG           286
TV-Y7        132
TV-G         118
TV-Y         112
NR            75
G             41
TV-Y7-FV       4
UR             3
NC-17          2
84 min         1
74 min         1
66 min         1
Name: rating, dtype: int64


 #From value counts we know that the most frequent value is TV-MA. So replacing the missing values with "TV-MA" 

In [7]:
netflix_title["rating"] = netflix_title["rating"].fillna("TV-MA")
netflix_title["rating"].value_counts() 
#From the value counts before replacing it with TV-MA, we know that there was 2117 values for TV-MA and now there is 2118, so 1 missing value in rating is replaced with mode.  

TV-MA       2118
TV-14       1457
R            795
TV-PG        538
PG-13        489
PG           286
TV-Y7        132
TV-G         118
TV-Y         112
NR            75
G             41
TV-Y7-FV       4
UR             3
NC-17          2
84 min         1
74 min         1
66 min         1
Name: rating, dtype: int64

### Question 4
Remove the rows which have empty value for the column “duration”. 

In [8]:
netflix_title = netflix_title.dropna(subset =['duration'])
netflix_title.isna().sum() #Confirming that the missing values for duration which was 3 was dropped. 

show_id           0
type              0
title             0
director          0
cast            473
country         422
date_added        0
release_year      0
rating            0
duration          0
listed_in         0
description       0
dtype: int64

### Question 5.
Take only movies from the dataset and extract the numeric values from the “duration” 
column. Save the extracted data in a new column “minutes”. For example, if a movie’s 
duration is “100 minutes”, then the value of its “minutes” should be 100.

In [9]:
duration_frame = netflix_title.loc[netflix_title["type"] =="Movie"] #first creating a new dataframe that has type listed as movie only.
print(duration_frame["type"].unique()) #Confirming that only unique values are left. 
duration_frame["minutes"] = duration_frame["duration"].str.split(" ").str[0].astype(int)
duration_frame["minutes"].head(10)


['Movie']


<ipython-input-9-77eaaa6cb0dc>:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  duration_frame["minutes"] = duration_frame["duration"].str.split(" ").str[0].astype(int)


### Question 6 : A 

Finding the number values > 50 and greater than 150 as outliers

In [10]:
outliers = (duration_frame["minutes"] <50) | (duration_frame["minutes"] > 150)
outliers.value_counts()




False    5480
True      460
Name: minutes, dtype: int64

There are 460 outlier values i.e who value is either less than 50 or more than 150. 

### Question 6 : B 

Filter out Outliers

In [11]:
duration_frame = duration_frame.loc[(duration_frame["minutes"]>50) & (duration_frame["minutes"]<150)]
duration_frame.head()


show_id   type                             title  \
0       s1  Movie              Dick Johnson Is Dead   
6       s7  Movie  My Little Pony: A New Generation   
7       s8  Movie                           Sankofa   
9      s10  Movie                      The Starling   
12     s13  Movie                      Je Suis Karl   

                         director  \
0                 Kirsten Johnson   
6   Robert Cullen, José Luis Ucha   
7                    Haile Gerima   
9                  Theodore Melfi   
12            Christian Schwochow   

                                                 cast  \
0                                                 NaN   
6   Vanessa Hudgens, Kimiko Glenn, James Marsden, ...   
7   Kofi Ghanaba, Oyafunmike Ogunlano, Alexandra D...   
9   Melissa McCarthy, Chris O'Dowd, Kevin Kline, T...   
12  Luna Wedler, Jannis Niewöhner, Milan Peschel, ...   

                                              country          date_added  \
0                                       United States  September 25, 2021   
6                                                 NaN  September 24, 2021   
7   United States, Ghana, Burkina Faso, United Kin...  September 24, 2021   
9                                       United States  September 24, 2021   
12                            Germany, Czech Republic  September 23, 2021   

    release_year rating duration  \
0           2020  PG-13   90 min   
6           2021     PG   91 min   
7           1993  TV-MA  125 min   
9           2021  PG-13  104 min   
12          2021  TV-MA  127 min   

                                           listed_in  \
0                                      Documentaries   
6                           Children & Family Movies   
7   Dramas, Independent Movies, International Movies   
9                                   Comedies, Dramas   
12                      Dramas, International Movies   

                                          description  minutes  
0   As her father nears the end of his life, filmm...       90  
6   Equestria's divided. But a bright-eyed hero be...       91  
7   On a photo shoot in Ghana, an American model s...      125  
9   A woman adjusting to life after a loss contend...      104  
12  After most of her family is murdered in a terr...      127

In [12]:
#Confirming that the outlier is filtered out from the dataset. 

In [13]:
max_minutes = duration_frame["minutes"].max()
min_minutes = duration_frame["minutes"].min()
print(max_minutes, min_minutes) 


149 51


So in the dataset above, we can see that thw maximum value for the minutes value is 149 and minimum value is 51 which means it dropped all the outliers from the duration_frame set.

### Question 7 

In [14]:
title = pd.read_csv("netflix_titles.csv")  #importing the dataset again 
original = pd.read_csv("netflix_originals.csv")  #importing another dataset


In [15]:
print(title.info(), original.info())


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8807 entries, 0 to 8806
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   show_id       8807 non-null   object
 1   type          8807 non-null   object
 2   title         8807 non-null   object
 3   director      6173 non-null   object
 4   cast          7982 non-null   object
 5   country       7976 non-null   object
 6   date_added    8797 non-null   object
 7   release_year  8807 non-null   int64 
 8   rating        8803 non-null   object
 9   duration      8804 non-null   object
 10  listed_in     8807 non-null   object
 11  description   8807 non-null   object
dtypes: int64(1), object(11)
memory usage: 825.8+ KB
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1517 entries, 0 to 1516
Data columns (total 10 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   titles       1517 non-null   object 
 1   years        1

In [16]:
## Titles needs to be case insenstive, strip of white spaces and removed punctuation

title["title"] = title["title"].str.lower().str.strip() #not case sensitive & strip from white spaces
punc_strip= lambda x: x.replace(r'[^\w\s]+', '')   #Function to remove the punctuations
title["title"] = title["title"].apply(punc_strip)   #applyign function to the title
#print(title["title"].unique())

#print(original.head())
original["titles"] = original["titles"].str.lower().str.strip()
original["titles"] = original["titles"].apply(punc_strip)



In [17]:
#original.head(20) #Confirming that functions worked on the original and title dataset
#print(title.head())

original = original.rename(columns={'titles': 'title', 'years': 'release_year'}) 
original.head()   #Confirming that the column name are changed.


title  release_year                 genres  imdb runtime  \
0  zumbo's just desserts          2016             Reality-TV   6.9  52 min   
1              zona rosa          2019                 Comedy   6.0     NaN   
2        young wallander          2020  Crime, Drama, Mystery   6.7     NaN   
3           you vs. wild          2019  Adventure, Reality-TV   6.7  20 min   
4                    you          2018  Crime, Drama, Romance   7.8  45 min   

                                         description  \
0  Amateur Australian chefs compete to impress pa...   
1                                         Add a Plot   
2  Follow recently graduated police officer Kurt ...   
3  In this interactive series, you'll make key de...   
4  A dangerously charming, intensely obsessive yo...   

                                               stars    votes     type  \
0           Gigi Falanga, Rachel Khoo, Adriano Zumbo    1,779  TV Show   
1  Ray Contreras, Pablo Morán, Manu Nna, Ana Juli...       33  TV Show   
2  Adam Pålsson, Leanne Best, Richard Dillane, El...    5,419  TV Show   
3                                        Bear Grylls    1,977  TV Show   
4  Penn Badgley, Victoria Pedretti, Ambyr Childer...  134,932  TV Show   

  original  
0  Netflix  
1  Netflix  
2  Netflix  
3  Netflix  
4  Netflix

In [18]:
merged_data = pd.merge(title, original, on = ["title", "release_year", "type"])
merged_data.head()  #COnfirming that the merged worked and there are no repition of vlaue 

show_id     type                 title         director  \
0      s1    Movie  dick johnson is dead  Kirsten Johnson   
1    s531  TV Show           richie rich              NaN   
2    s554  TV Show                mortel              NaN   
3    s674  TV Show     no time for shame              NaN   
4    s681  TV Show        the last dance              NaN   

                                                cast        country  \
0                                                NaN  United States   
1  Jake Brennan, Joshua Carlon, Lauren Taylor, Je...  United States   
2  Carl Malapa, Nemo Schiffman, Manon Bresch, Cor...         France   
3                                   Santiago Artemis      Argentina   
4      Michael Jordan, Scottie Pippen, Dennis Rodman  United States   

           date_added  release_year rating   duration  \
0  September 25, 2021          2020  PG-13     90 min   
1        July 6, 2021          2015   TV-G  2 Seasons   
2        July 2, 2021          2019  TV-MA  2 Seasons   
3       June 19, 2021          2019  TV-MA   1 Season   
4       June 19, 2021          2020  TV-MA   1 Season   

                                           listed_in  \
0                                      Documentaries   
1                              Kids' TV, TV Comedies   
2  Crime TV Shows, International TV Shows, TV Dramas   
3  International TV Shows, Reality TV, Spanish-La...   
4                                         Docuseries   

                                       description_x  \
0  As her father nears the end of his life, filmm...   
1  After turning his veggies into green energy, R...   
2  After making a deal with a supernatural figure...   
3  Follow Santiago Artemis, a Buenos Aires fashio...   
4  This docuseries gives a definitive account of ...   

                            genres  imdb  runtime  \
0               Documentary, Drama   7.6   89 min   
1                           Comedy   3.2   22 min   
2                   Horror, Sci-Fi   6.5   40 min   
3                       Reality-TV   5.5      NaN   
4  Documentary, Biography, History   9.2  491 min   

                                       description_y  \
0  A daughter helps her father prepare for the en...   
1  Richie Rich is just a normal kid, except he ha...   
2  A group of French teenagers are bound together...   
3  The story of Argentinian fashion designer Sant...   
4  Charting the rise of the 1990's Chicago Bulls,...   

                                               stars   votes original  
0  Kirsten Johnson, Michael Hilow, Ana Hoffman, D...   1,426  Netflix  
1  Jake Brennan, Lauren Taylor, Joshua Carlon, Je...     790  Netflix  
2  Carl Malapa, Némo Schiffman, Manon Bresch, Cor...     748  Netflix  
3                    Edwin Salinas, Santiago Artemis      67  Netflix  
4  Phil Jackson, Michael Jordan, David Aldridge, ...  71,228  Netflix